core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java

```diff
@@ -259,9 +260,25 @@ public UfsStatus[] listStatus(String path, ListOptions options) throws IOExcepti
     return statuses;
   }
 
+  /**
+   * @param fileInfo the FileInfo of this file. Cached pages are identified by PageId
+   * @return true at this moment
+   */
+  @Override
+  public boolean invalidateCachedFile(FileInfo fileInfo) {
+    long pages = (fileInfo.getLength() + mPageSize - 1) / mPageSize;
+    for (long i = 0; i < pages; i++) {
+      FileId file = FileId.of(new AlluxioURI(fileInfo.getUfsPath()).hash());
+      PageId page = new PageId(file.toString(), i);
+      mCacheManager.delete(page);
```

Comment:
This is deleting every possible page of the file, based on the file length, regardless of whether it is indeed being cached in the page store. Some pages not being cached will fail to be deleted, and secretly increments the error counter: https://github.com/Alluxio/alluxio/blob/317523a641af5f5831a7e9e9937459c6342b2898/core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java#L520-L523
I think the invalidation should only be done on pages that already exist in the page store.

Reply:
While pages of the given file are being deleted, if the client concurrently reads in new , up-to-date pages from UFS, will they be deleted accidentally as well?

Reply:
> This is deleting every possible page of the file, based on the file length, regardless of whether it is indeed being cached in the page store. Some pages not being cached will fail to be deleted, and secretly increments the error counter:
> 
> https://github.com/Alluxio/alluxio/blob/317523a641af5f5831a7e9e9937459c6342b2898/core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java#L520-L523
> 
> 
> I think the invalidation should only be done on pages that already exist in the page store.

Such `error` is not a real error. So we may change the LOG level to debug/trace.

If we want to know the exact pages that are already cached, we need to iterate and search the existing cache. Checking the existence and then delete will have the same performance as delete()?

Reply:
> While pages of the given file are being deleted, if the client concurrently reads in new , up-to-date pages from UFS, will they be deleted accidentally as well?


This will be a race condition. Either we need to have some lock, or to tolerate such race condition.
Currently, Dora is for read-only or (mostly read) case. So such invalidation is just for rare case in which the users know that we have updated some files. So they can invalidate the cached files __before__ reading them.

Reply:
> If we want to know the exact pages that are already cached, we need to iterate and search the existing cache. 

Or we need to establish a mapping between the files and the pages of each file. So that we know all pages that belong to a certain file. We did the similar thing in the paged block store:

https://github.com/Alluxio/alluxio/blob/669f80e612f2086d1ded170cfc09cdddc3ffbc2c/core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java#L515

Reply:
> > If we want to know the exact pages that are already cached, we need to iterate and search the existing cache.
> 
> Or we need to establish a mapping between the files and the pages of each file. So that we know all pages that belong to a certain file without the need to iterate all page files. We did the similar thing in the paged block store:
> 
> https://github.com/Alluxio/alluxio/blob/669f80e612f2086d1ded170cfc09cdddc3ffbc2c/core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java#L515

Yes, maybe we can add an interface to CacheManager,
CacheManager::deletePages(FileId fileld);

This will delete all cached pages for this File.
This function may use the in-memory metadata to determine the cached pages of this file.